<a href="https://colab.research.google.com/github/Kanareiika/DeepLearningSchool/blob/master/03.%20Pandas%2C%20Matplotlib%2C%20ML%20basics/%5Bseminar%5Dpandas.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

<h2 style="text-align: center;"><b>Библиотека <a href="http://pandas.pydata.org/">pandas</a></b></h2>

Библиотека `pandas` активно используется в современном data science для работы с данными, которые могут быть представлены в виде таблиц (а это очень, очень большая часть данных)

`pandas` есть в пакете Anaconda, но если вдруг у Вас её по каким-то причинам нет, то можно установить, раскомментировав следующую команду:

In [0]:
#!pip install pandas

In [0]:
import pandas as pd

Нужны данные. Про AppStore слышали? :)

### Скачать данные: https://www.kaggle.com/ramamet4/app-store-apple-data-set-10k-apps

Нас интересует файл `AppStore.csv`.

Кстати, `.csv` (Comma Separated Values) - наверное, самый частый формат данных в современном data science. По сути, это то же самое, что и `.xls` (и `.xlsx`), то есть таблица. 

Итак, *чтение файла с данными*:

In [0]:
data = pd.read_csv('https://media.githubusercontent.com/media/DLSchool/dlschool/master/03.%20Pandas%2C%20Matplotlib%2C%20ML%20basics/app-store-apple-data-set-10k-apps/AppleStore.csv')

Посмотрим, что же такое переменная `data`:

In [3]:
data

Unnamed: 0          id  ... lang.num  vpp_lic
0              1   281656475  ...       10        1
1              2   281796108  ...       23        1
2              3   281940292  ...        3        1
3              4   282614216  ...        9        1
4              5   282935706  ...       45        1
5              6   283619399  ...        1        1
6              7   283646709  ...       19        1
7              8   284035177  ...        1        1
8              9   284666222  ...        1        1
9             10   284736660  ...       10        1
10            11   284791396  ...       11        1
11            12   284815117  ...        6        1
12            13   284815942  ...       33        1
13            14   284847138  ...        2        1
14            15   284862767  ...        2        1
15            16   284876795  ...       26        1
16            17   284882215  ...       29        1
17            18   284910350  ...       18        1
18            20   284993459  ...       16        1
19            21   285005463  ...        1        1
20            22   285946052  ...        2        1
21            23   285994151  ...        1        1
22            24   286058814  ...        6        1
23            25   286070473  ...        3        1
24            27   286799607  ...       10        1
25            28   286906691  ...       11        1
26            29   286911400  ...        1        1
27            31   288113403  ...       23        1
28            32   288120394  ...        1        1
29            33   288419283  ...        1        1
...          ...         ...  ...      ...      ...
7167       10995  1182265441  ...        1        1
7168       10998  1182331762  ...        1        1
7169       11002  1182568288  ...       31        1
7170       11010  1183234072  ...        1        1
7171       11013  1183260922  ...        1        1
7172       11016  1183548754  ...        1        1
7173       11019  1183709176  ...        1        1
7174       11022  1183856228  ...        1        1
7175       11024  1183986102  ...       25        1
7176       11027  1184711626  ...        1        1
7177       11031  1184800011  ...        1        1
7178       11033  1185209084  ...        1        1
7179       11035  1185328193  ...        1        1
7180       11036  1185365336  ...        1        1
7181       11038  1185428381  ...        0        1
7182       11040  1185538497  ...        1        1
7183       11041  1185580782  ...        1        1
7184       11042  1185731859  ...        0        1
7185       11043  1185777521  ...        1        1
7186       11050  1186108496  ...        1        1
7187       11051  1186126548  ...        2        1
7188       11060  1186384912  ...        1        1
7189       11074  1187128255  ...        1        1
7190       11077  1187279979  ...        3        1
7191       11079  1187282363  ...        1        1
7192       11081  1187617475  ...        1        1
7193       11082  1187682390  ...        1        1
7194       11087  1187779532  ...        1        1
7195       11089  1187838770  ...        2        1
7196       11097  1188375727  ...        2        1

[7197 rows x 17 columns]

# New Section

In [4]:
type(data)

pandas.core.frame.DataFrame

Тип == `pandas.core.frame.DataFrame`, обычно говорят просто "датафрейм", то есть "кусок данных".

Можно вывести информацию по датафрейму:

In [5]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7197 entries, 0 to 7196
Data columns (total 17 columns):
Unnamed: 0          7197 non-null int64
id                  7197 non-null int64
track_name          7197 non-null object
size_bytes          7197 non-null int64
currency            7197 non-null object
price               7197 non-null float64
rating_count_tot    7197 non-null int64
rating_count_ver    7197 non-null int64
user_rating         7197 non-null float64
user_rating_ver     7197 non-null float64
ver                 7197 non-null object
cont_rating         7197 non-null object
prime_genre         7197 non-null object
sup_devices.num     7197 non-null int64
ipadSc_urls.num     7197 non-null int64
lang.num            7197 non-null int64
vpp_lic             7197 non-null int64
dtypes: float64(3), int64(9), object(5)
memory usage: 955.9+ KB


И более статистически значимую информацию:

In [6]:
data.describe()

Unnamed: 0            id  ...     lang.num      vpp_lic
count   7197.000000  7.197000e+03  ...  7197.000000  7197.000000
mean    4759.069612  8.631310e+08  ...     5.434903     0.993053
std     3093.625213  2.712368e+08  ...     7.919593     0.083066
min        1.000000  2.816565e+08  ...     0.000000     0.000000
25%     2090.000000  6.000937e+08  ...     1.000000     1.000000
50%     4380.000000  9.781482e+08  ...     1.000000     1.000000
75%     7223.000000  1.082310e+09  ...     8.000000     1.000000
max    11097.000000  1.188376e+09  ...    75.000000     1.000000

[8 rows x 12 columns]

Итак, наши данные -- это какая-то **информация про приложения в AppStore**. Давайте посмотрим, что тут есть:

In [7]:
# все столбцы датафрейма (названия)
data.columns

Index(['Unnamed: 0', 'id', 'track_name', 'size_bytes', 'currency', 'price',
       'rating_count_tot', 'rating_count_ver', 'user_rating',
       'user_rating_ver', 'ver', 'cont_rating', 'prime_genre',
       'sup_devices.num', 'ipadSc_urls.num', 'lang.num', 'vpp_lic'],
      dtype='object')

---

#### Задание 0

О чём данные? (судя по столбцам)

---

Все значения датафрейма:

In [8]:
# одна строка == описание одного объекта, в данном случае -- приложения из AppStore
data.values

array([[1, 281656475, 'PAC-MAN Premium', ..., 5, 10, 1],
       [2, 281796108, 'Evernote - stay organized', ..., 5, 23, 1],
       [3, 281940292, 'WeatherBug - Local Weather, Radar, Maps, Alerts',
        ..., 5, 3, 1],
       ...,
       [11087, 1187779532, 'Bret Michaels Emojis + Lyric Keyboard', ...,
        1, 1, 1],
       [11089, 1187838770, 'VR Roller Coaster World - Virtual Reality',
        ..., 0, 2, 1],
       [11097, 1188375727, 'Escape the Sweet Shop Series', ..., 0, 2, 1]],
      dtype=object)

In [9]:
data.values[0]

array([1, 281656475, 'PAC-MAN Premium', 100788224, 'USD', 3.99, 21292, 26,
       4.0, 4.5, '6.3.5', '4+', 'Games', 38, 5, 10, 1], dtype=object)

In [10]:
data.shape

(7197, 17)

-- то есть 7197 строк и 17 столбцов.

Важно уметь обращаться к определённой строке или столбцу dataframe'а. Индексирование очень похоже на numpy, однако есть свои тонкости:

* Получить весь столбец (в даном случае столбец 'track_name'):

In [11]:
data['track_name']

0                                         PAC-MAN Premium
1                               Evernote - stay organized
2         WeatherBug - Local Weather, Radar, Maps, Alerts
3       eBay: Best App to Buy, Sell, Save! Online Shop...
4                                                   Bible
5                                        Shanghai Mahjong
6                  PayPal - Send and request money safely
7                                 Pandora - Music & Radio
8                             PCalc - The Best Calculator
9                                             Ms. PAC-MAN
10                              Solitaire by MobilityWare
11                                       SCRABBLE Premium
12                   Google – Search made just for mobile
13                       Bank of America - Mobile Banking
14                                               FreeCell
15                 TripAdvisor Hotels Flights Restaurants
16                                               Facebook
17         Yel

In [12]:
type(data['track_name'])

pandas.core.series.Series

In [13]:
data['track_name'][0]

'PAC-MAN Premium'

`pandas.core.series.Series` -- это тип подвыборки датафрейма. С этим типом работают реже, но стоит знать, что он есть.

* Получить определённую строку:

In [14]:
data.iloc[10]

Unnamed: 0                                 11
id                                  284791396
track_name          Solitaire by MobilityWare
size_bytes                           49618944
currency                                  USD
price                                    4.99
rating_count_tot                        76720
rating_count_ver                         4017
user_rating                               4.5
user_rating_ver                           4.5
ver                                    4.10.1
cont_rating                                4+
prime_genre                             Games
sup_devices.num                            38
ipadSc_urls.num                             4
lang.num                                   11
vpp_lic                                     1
Name: 10, dtype: object

Более продвинутое и полезное индексирование:

In [15]:
data.iloc[5, 10]

'1.8'

-- `data.iloc[i, j]`, где `i` -- номер строки, `j` -- номер столбца

Почти все типы индексирования:

* к строкам по числовому индексу, к столбцам -- по имени

In [16]:
data.loc[[0, 1, 2], ['track_name', 'id']]

track_name         id
0                                  PAC-MAN Premium  281656475
1                        Evernote - stay organized  281796108
2  WeatherBug - Local Weather, Radar, Maps, Alerts  281940292

* к строкам и к столбцам по числовому индексу

In [17]:
data.iloc[[1, 3, 5], [2, 1]]

track_name         id
1                          Evernote - stay organized  281796108
3  eBay: Best App to Buy, Sell, Save! Online Shop...  282614216
5                                   Shanghai Mahjong  283619399

* поддерживает и то, и то, но **запрещён**:

In [18]:
data.ix[[2,1,0], ['track_name', 'id']]

/usr/local/lib/python3.6/dist-packages/ipykernel_launcher.py:1: DeprecationWarning: 
.ix is deprecated. Please use
.loc for label based indexing or
.iloc for positional indexing

See the documentation here:
http://pandas.pydata.org/pandas-docs/stable/indexing.html#ix-indexer-is-deprecated
  """Entry point for launching an IPython kernel.


track_name         id
2  WeatherBug - Local Weather, Radar, Maps, Alerts  281940292
1                        Evernote - stay organized  281796108
0                                  PAC-MAN Premium  281656475

In [19]:
data.ix[[0,1,2], [2, 1]]

/usr/local/lib/python3.6/dist-packages/ipykernel_launcher.py:1: DeprecationWarning: 
.ix is deprecated. Please use
.loc for label based indexing or
.iloc for positional indexing

See the documentation here:
http://pandas.pydata.org/pandas-docs/stable/indexing.html#ix-indexer-is-deprecated
  """Entry point for launching an IPython kernel.


track_name         id
0                                  PAC-MAN Premium  281656475
1                        Evernote - stay organized  281796108
2  WeatherBug - Local Weather, Radar, Maps, Alerts  281940292

Точно также дело обстоит со **срезами**:

In [20]:
data.iloc[5:100, 0:5]

Unnamed: 0         id  ... size_bytes  currency
5            6  283619399  ...   10485713       USD
6            7  283646709  ...  227795968       USD
7            8  284035177  ...  130242560       USD
8            9  284666222  ...   49250304       USD
9           10  284736660  ...   70023168       USD
10          11  284791396  ...   49618944       USD
11          12  284815117  ...  227547136       USD
12          13  284815942  ...  179979264       USD
13          14  284847138  ...  160925696       USD
14          15  284862767  ...   55153664       USD
15          16  284876795  ...  207907840       USD
16          17  284882215  ...  389879808       USD
17          18  284910350  ...  167407616       USD
18          20  284993459  ...  147093504       USD
19          21  285005463  ...   10735026       USD
20          22  285946052  ...   70707916       USD
21          23  285994151  ...    6169600       USD
22          24  286058814  ...  130583552       USD
23          25  286070473  ...   71203840       USD
24          27  286799607  ...   11423008       USD
25          28  286906691  ...  188017664       USD
26          29  286911400  ...    4765696       USD
27          31  288113403  ...  287933440       USD
28          32  288120394  ...    4263936       USD
29          33  288419283  ...  172772352       USD
30          34  288429040  ...  273844224       USD
31          35  289084315  ...   40216576       USD
32          36  289446241  ...    2386944       USD
33          37  289523017  ...  105431040       USD
34          39  289894882  ...   44129280       USD
..         ...        ...  ...        ...       ...
70          80  299853944  ...  115143680       USD
71          81  299949744  ...   56481792       USD
72          82  300048137  ...  181941248       USD
73          83  300238550  ...  162891776       USD
74          84  300255638  ...   98108416       USD
75          85  300265786  ...   54088704       USD
76          86  300590611  ...   50155520       USD
77          87  300704847  ...   23936000       USD
78          89  301387274  ...  145293312       USD
79          90  301521403  ...  140367872       USD
80          91  301724680  ...  281244672       USD
81          92  302049354  ...   66695168       USD
82          93  302584613  ...  169747456       USD
83          95  303058767  ...   41691136       USD
84          96  303191318  ...  122886144       USD
85          97  303337299  ...   55803904       USD
86          99  303849934  ...  188956672       USD
87         100  304731501  ...   22185984       USD
88         101  304770340  ...   66160640       USD
89         102  304871603  ...   28135424       USD
90         103  304871622  ...   27962441       USD
91         104  304878510  ...  133238784       USD
92         105  305343404  ...  151573504       USD
93         106  305557780  ...   27069440       USD
94         107  305939712  ...  101259264       USD
95         108  306169895  ...  168192000       USD
96         109  306257910  ...   17485824       USD
97         110  306375551  ...    5234688       USD
98         111  306468004  ...  237913088       USD
99         112  306561234  ...  233513984       USD

[95 rows x 5 columns]

Но обычно нужно уметь отвечать на более осмысленные вопросы, например:

* сколько приложений имеют рейтинг юзеров >= 4?

In [22]:
len(data[data['user_rating'] >= 4])

4781

* какие есть валюты в столбце currency?

In [23]:
import numpy as np
np.unique(data['currency'])

array(['USD'], dtype=object)

* сколько приложений жанра Games? А тех, которые Games и рейтинг >= 4?

In [24]:
# 1
len(data[data['prime_genre'] == 'Games'])

3862

In [25]:
# 2
len(data[(data['user_rating'] >= 4) & (data['prime_genre'] == 'Games')])

2868

Оператор "ИЛИ" аналогично &, только он через |

In [26]:
len(data[(data['user_rating'] >= 4) | (data['prime_genre'] == 'Games')])

5775

---

#### Задание 1

Получите ответ на следующие вопросы (с помощью pandas):

1). Сколько всего жанров и какие они?  
2). Сколько суммарно байт весят все приложения жанра Finance?  
3). Сколько всего бесплатных приложений, у которых рейтинг юзеров больше трёх?  
4). Есть ли приложения, которые стоят больше 100 долларов?

In [27]:
print('Сколько всего жанров и какие они?', np.unique(data['prime_genre']))
print('Сколько суммарно байт весят все приложения жанра Finance?', np.sum(  (data[data['prime_genre'] == 'Finance'])  ['size_bytes']  )/(2**30), 'Gb') #командой data[data['prime_genre'] == 'Finance'] мы берем кусочек data только с финансовыми приложениями - выглядит громоздко
print('Сколько всего бесплатных приложений, у которых рейтинг юзеров больше трёх?', len(  data[(data['user_rating'] >= 3) & (data['price'] == 0)]  ))
print('Есть ли приложения, которые стоят больше 100 долларов?')
if (len (data[data['price'] >= 100]) > 0):
    print ('Да')
else:
    print('Нет')

Сколько всего жанров и какие они? ['Book' 'Business' 'Catalogs' 'Education' 'Entertainment' 'Finance'
 'Food & Drink' 'Games' 'Health & Fitness' 'Lifestyle' 'Medical' 'Music'
 'Navigation' 'News' 'Photo & Video' 'Productivity' 'Reference' 'Shopping'
 'Social Networking' 'Sports' 'Travel' 'Utilities' 'Weather']
Сколько суммарно байт весят все приложения жанра Finance? 7.577733477577567 Gb
Сколько всего бесплатных приложений, у которых рейтинг юзеров больше трёх? 3160
Есть ли приложения, которые стоят больше 100 долларов?
Да


---

Часто надо сделать что-то такое:

In [0]:
import numpy as np

In [29]:
np.array(data['size_bytes'])

array([100788224, 158578688, 100524032, ..., 111322112,  97235968,
        90898432])

То есть взять столбец из таблицы и засунуть его в np.array(). Это можно сделать и просто обратившись к .values, ведь его тип и есть np.array:

In [30]:
type(data['size_bytes'].values)

numpy.ndarray

Больше про pandas можно найти по этом полезным ссылкам:

* Официальные туториалы: http://pandas.pydata.org/pandas-docs/stable/tutorials.html

* Статья на Хабре от OpenDataScience сообщества: https://habr.com/company/ods/blog/322626/

* Подробный гайд: https://media.readthedocs.org/pdf/pandasguide/latest/pandasguide.pdf

Главное в работе с новыми библиотеками -- не бояться тыкать в разные функции, смотреть типы возвращаемых объектов и активно пользоваться Яндексом, а ещё лучше понимать всё из docstring'а (`Shift+Tab` при нахождении курсора внутри скобок функции).